### PR TITLE
Web app

### PR DESCRIPTION
**Pull request by KwassiSenam**
<h1>What has been done</h1>
The present PR gathers the latest updates that have been made to the frontend.  These changes include:
<br />
<h2>1. Update of new pages </h2>
The following pages have been modified:

- About : the table of contents has been reduced (there is no more **Programme** and **Notice d'utilisation** parts) .

- Run : the available datasets are retrieved using a get request.
- Logs :  the review option has been implemented
- Operation page : all the logics and add a route graph
- Results page : all the logics and add a route graph

<h2>2. Implementation of new pages </h2>
The review page has been implemented. This page allows the user to review an old resolution.

<h2>Coverage</h2>

image

image

### CODE REVIEW

**Comment by galyfray** [comment]
Please merge master into your branch, there are some conflicts that must be resolved.

File: src/web_app/src/components/appBar.tsx [hunk 1]
```diff
@@ -17,7 +17,7 @@ export function AppbarStyle() {
                 <Typography variant="h4" color="inherit" noWrap sx={{
                     fontWeight: "bold", flexGrow: 1, ml: 5
                 }}>
-            Semina
+            Semita
```
Comment (galyfray) [inline src/web_app/src/components/appBar.tsx:20]: didn't we agree on Semina ?

File: src/web_app/src/pages/drl_hyperParams.tsx [hunk 2]
```diff
@@ -10,13 +10,37 @@ import Grid from "@mui/material/Grid";
...
 import Stack from "@mui/material/Stack";
 import Button from "@mui/material/Button";
-import * as Types from "../types/data";
 import axios from "axios";
 import Backdrop from "@mui/material/Backdrop";
-import sleep from "../config/sleep_funct";
 import CircularProgress from "@mui/material/CircularProgress";
 import FormControlLabel from "@mui/material/FormControlLabel";
 import Checkbox from "@mui/material/Checkbox";
+const datasets = [
```
Comment (galyfray) [inline src/web_app/src/pages/drl_hyperParams.tsx:18]: This const value should not be here anymore

File: src/web_app/src/pages/ga_hyperParams.tsx [hunk 3]
```diff
@@ -10,13 +10,38 @@ import Grid from "@mui/material/Grid";
...
-import * as Types from "../types/data";
-import axios from "axios";
 import Backdrop from "@mui/material/Backdrop";
-import sleep from "../config/sleep_funct";
 import CircularProgress from "@mui/material/CircularProgress";
 import FormControlLabel from "@mui/material/FormControlLabel";
 import Checkbox from "@mui/material/Checkbox";
+import axios from "axios";
+
+const datasets = [
```
Comment (galyfray) [inline src/web_app/src/pages/ga_hyperParams.tsx:19]: Same here, this const value should not be here anymore

File: src/web_app/src/pages/ga_hyperParams.tsx [hunk 4]
```diff
@@ -87,20 +95,22 @@ export function GaHyperParamsPage() {
                 "momentum"     : 0.2
```
Comment (galyfray) [inline src/web_app/src/pages/ga_hyperParams.tsx:95]: why is there a momentum and a learning rate in the GA hyper parameters ?

File: src/web_app/src/pages/ga_hyperParams.tsx [hunk 5]
```diff
@@ -163,13 +175,13 @@ export function GaHyperParamsPage() {
...
-                            defaultValue={param.params.seed}
+                            defaultValue={seed}
                             onChange={e => setSeed(parseInt(e.target.value))}>
                         </TextField>
                     </Grid>
                     <Grid item xs={6}>
                         <TextField id="filled-basic" variant="filled"
-                            defaultValue={param.params.mutation_rate}
+                            defaultValue={mutation_rate}
                             onChange={e => setMutation_rate(parseInt(e.target.value))}>
```
Comment (galyfray) [inline src/web_app/src/pages/ga_hyperParams.tsx:185]: Mutation rate is a float not an Int this will cause problems ...

File: src/web_app/src/pages/operation.tsx [hunk 6]
```diff
@@ -4,202 +4,212 @@ import CssBaseline from "@mui/material/CssBaseline";
...
+import Badge from "@mui/material/Badge";
+import Stack from "@mui/material/Stack";
+import {
+    getMaxX,
+    getMaxY,
+    getMinFitness, getMinX, getMinY, getRandomColor
+} from "../config/utils";
+import Avatar from "@mui/material/Avatar";
+
+const datasets = [
```
Comment (galyfray) [inline src/web_app/src/pages/operation.tsx:24]: here again ...

File: src/web_app/src/pages/result.tsx [hunk 7]
```diff
@@ -9,75 +9,190 @@ import {AppbarStyle} from "../components/appBar";
...
-    LineChart, Line, XAxis, YAxis, CartesianGrid, Tooltip, Legend
+    LineChart, Line, XAxis, YAxis, CartesianGrid, Tooltip, LabelList
 } from "recharts";
-import * as d3Types from "../types/d3Types";
 import * as Types from "../types/data";
+import {
+    getMaxX, getMaxY, getMinX, getMinY, getRandomColor
+} from "../config/utils";
 
+const datasets = [
```
Comment (galyfray) [inline src/web_app/src/pages/result.tsx:19]: how many of them are they ?

File: src/web_app/src/pages/review.tsx [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+                >
+                    Evolution de la fitness au cours des générations
+                </Typography>
+
+                <Grid container alignItems="center" spacing={2}>
+                    <Grid item xs={6} sx={{mt: 4}}>
+                        <LineChart width={500} height={300}
+                            data={plotdata.slice(1, plotdata.length)}
+                        >
+                            <Line type="monotone" dataKey="fitness" stroke="#82ca9d" strokeWidth={2} />
```
Comment (galyfray) [inline src/web_app/src/pages/review.tsx:244]: seems like this one is animated while the other ones are not

File: src/web_app/src/pages/run.tsx [hunk 9]
```diff
@@ -206,6 +156,11 @@ export function RunPage() {
                 Suivant
                     </Button>
                 </Stack>
+                { open &&
+            <Alert onClose={handleClose} severity="warning" sx={{width: "100%"}}>
+            Le solveur est actuellement entrain de tourner. En cliquant sur
+            Suivant, vous redémarrez une nouvelle résolution.
```
Comment (galyfray) [inline src/web_app/src/pages/run.tsx:162]: well if you try to run a new resolution the server will answer as stated in the documentation but won't start any new resolution unless the previous one is done ^^.

**Review by galyfray — CHANGES_REQUESTED** [review]
There are some minors things to change here and there but nothing too important

File: src/web_app/src/pages/ga_hyperParams.tsx (hunk 5 above)
Comment (KwassiSenam) [reply src/web_app/src/pages/ga_hyperParams.tsx:185]: All numbers in TypeScript are either floating point values or integers.

File: src/web_app/src/components/appBar.tsx (hunk 1 above)
Comment (KwassiSenam) [reply src/web_app/src/components/appBar.tsx:20]: I am not sure. Ok I will change this.

File: src/web_app/src/pages/drl_hyperParams.tsx (hunk 2 above)
Comment (KwassiSenam) [reply src/web_app/src/pages/drl_hyperParams.tsx:18]: I use it to get the corresponding dataset by taking as argument the dataset number in the url.

**Review by Axelvel — CHANGES_REQUESTED** [review]
Seems good to me overall, but you need to change the urls of the endpoints.

File: src/web_app/src/pages/drl_hyperParams.tsx (hunk 2 above)
Comment (galyfray) [reply src/web_app/src/pages/drl_hyperParams.tsx:18]: This can be a problem for future addition of new datasets, in any case, this list is duplicated in many files, can you at least factor the list out to reduce the problem ?

File: src/web_app/src/pages/ga_hyperParams.tsx (hunk 5 above)
Comment (galyfray) [reply src/web_app/src/pages/ga_hyperParams.tsx:185]: Yes, but by using `parseInt` here, you are transforming all float values between 0 and 1 to either 0 or 1. This is quite a big problem for a mutation rate ...

File: src/web_app/src/config/utils.ts [hunk 10]
```diff
@@ -0,0 +1,57 @@
+import {Link} from "../types/d3Types";
+import * as Types from "../types/data";
+import * as d3Types from "../types/d3Types";
+
+export function getRandomColor() {
+    return "#" + Math.floor(Math.random() * 16777215).toString(16)
```
Comment (galyfray) [inline src/web_app/src/config/utils.ts:6]: Some colors have a terrible contrast with the white background, if I well understand the code this could be fixed by changing this function, would it be a hard fix  ?

File: src/web_app/src/pages/ga_hyperParams.tsx [hunk 11]
```diff
@@ -87,20 +95,22 @@ export function GaHyperParamsPage() {
...
+        })
+            .then(response => {
+                console.log(response);
                 setOpen(true);
-                await sleep(5000);
+                setTimeout(() => {
                 // eslint-disable-next-line @typescript-eslint/no-unsafe-call
-                navigate(url + "operation");
-                setOpen(false);
+                    navigate(url + "operation", {state: {nb_epochs: inter.params.nb_epochs}}); //Comment this because the operation page not ready yet
```
Comment (galyfray) [inline src/web_app/src/pages/ga_hyperParams.tsx:111]: I think the comment can be removed ^^

File: src/web_app/src/pages/operation.tsx [hunk 12]
```diff
@@ -4,202 +4,212 @@ import CssBaseline from "@mui/material/CssBaseline";
...
-                    setMethod_str("Deep Reinforcement Learning");
-                }
-            });
-    }, [
-        dataset_choice,
-        method_choice
-    ]);
+    useEffect(() => {
+        if (datasets[parseInt(dataset_choice)]) {
+            setNb_cars(datasets[parseInt(dataset_choice)].split("-")[2][1]);
```
Comment (galyfray) [inline src/web_app/src/pages/operation.tsx:114]: if you have say 10 cars, this part of the code will only register 1 car.

**Review by galyfray — CHANGES_REQUESTED** [review]

File: src/web_app/src/pages/drl_hyperParams.tsx (hunk 2 above)
Comment (galyfray) [reply src/web_app/src/pages/drl_hyperParams.tsx:18]: > You can fetch the list of datasets using the `localhost:5001/benchmarks/` endpoint instead.

This is not possible, the list is used here as the benchmark is passed from page to page via an integer representing the index of the benchmark in the list. The problem is : the server does not guarantee the order of the benchmarks when responding to this route.

File: src/web_app/src/pages/operation.tsx (hunk 12 above)
Comment (KwassiSenam) [reply src/web_app/src/pages/operation.tsx:114]: oh yes. Good point ! I'll fix this.

File: src/web_app/src/pages/ga_hyperParams.tsx (hunk 4 above)
Comment (KwassiSenam) [reply src/web_app/src/pages/ga_hyperParams.tsx:95]: They will not be used. As you said, I send you all and according to the method type you select the corresponding hyperparameters.

File: src/web_app/src/pages/ga_hyperParams.tsx (hunk 5 above)
Comment (KwassiSenam) [reply src/web_app/src/pages/ga_hyperParams.tsx:185]: Ok I didn't understand the problem, I wil change parseInt to parseFloat.

File: src/web_app/src/components/log_row.tsx [hunk 13]
```diff
@@ -11,15 +12,25 @@ import Typography from "@mui/material/Typography";
...
 import TableHead from "@mui/material/TableHead";
 import TableBody from "@mui/material/TableBody";
+import PlayCircleFilledIcon from "@mui/icons-material/PlayCircleFilled";
+import Button from "@mui/material/Button";
+import {useLocation, useNavigate} from "react-router-dom";
 
 
 export default function Row(props: { row: Types.Log }) {
     const {row} = props;
+    const url = useLocation().pathname;
```
Comment (galyfray) [inline src/web_app/src/components/log_row.tsx:22]: This line makes a test fails. The test is the `log_page.test.tsx` and among the 400 pages jest has produced to explain to me that the test failed, I found this line `useLocation() may be used only in the context of a <Router> component.` I hope it will be useful. If you need the complete logs, I can send them to you.

**Review by galyfray — CHANGES_REQUESTED** [review]

File: src/web_app/src/components/log_row.tsx (hunk 13 above)
Comment (galyfray) [reply src/web_app/src/components/log_row.tsx:22]: I fixed it myself.

**Review by galyfray — APPROVED** [review]

**Review by galyfray — APPROVED** [review]

**Review by Axelvel — APPROVED** [review]
